lib/Dialect/HW/Transforms/FlattenModules.cpp

```diff
@@ -88,19 +148,89 @@ struct PrefixingInliner : public InlinerInterface {
     }
   }
 
+  void updateInnerSymbols(Operation *op) const {
+    // Rename inner symbols to avoid conflicts
+    if (auto innerSymAttr =
+            op->getAttrOfType<hw::InnerSymAttr>(innerSymAttrName)) {
+      StringAttr symName = innerSymAttr.getSymName();
+      auto it = symMapping->find(symName);
+      if (it != symMapping->end())
+        op->setAttr(innerSymAttrName, hw::InnerSymAttr::get(it->second));
+    }
+
+    // Apply attribute replacements for InnerRefAttr
+    replacer->replaceElementsIn(op);
+  }
+
   bool allowSingleBlockOptimization(
       iterator_range<Region::iterator> inlinedBlocks) const final {
     return true;
   }
 };
 } // namespace
 
+bool FlattenModulesPass::shouldInline(HWModuleOp module,
+                                      igraph::InstanceGraphNode *instanceNode,
+                                      size_t bodySize) {
+  // If inlineAll is enabled, inline everything (default behavior)
+  if (this->inlineAll)
+    return true;
+
+  // Check whether the module should be inlined based on heuristics.
+  bool isEmpty = bodySize == 1;
+  bool hasNoOutputs = module.getNumOutputPorts() == 0;
+  bool hasOneUse = instanceNode->getNumUses() == 1;
+  bool hasState = !module.getOps<seq::FirRegOp>().empty();
```

Comment:
Nit: I would move these closer to their usage sites as opposed to declaring them up top.